### PR TITLE
Multifile sources assembly

### LINKED ISSUE #8171: Legacy Assembly refers to comments in the code
## Description

On compiling this contract on remix:
```
/**
 * Foobar
 */
 
pragma solidity 0.6.1;

contract C1 {
    function f2() pure external returns (bool) {
        require(false);
        return true;
    }    

    function f1() view external {
        C1(this).f2();
    }
}
```
in compilation details, assembly is pointing comments:

```
.code
  PUSH 80			contract C1 {\n    function f2...
  PUSH 40			contract C1 {\n    function f2...
  MSTORE 			contract C1 {\n    function f2...
  CALLVALUE 			contract C1 {\n    function f2...
  DUP1 			oobar\n */
  ISZERO 			* 
  PUSH [tag] 1			* 
  JUMPI 			* 
  PUSH 0			i
  DUP1 			s
  REVERT 			pragma solid
tag 1			* 
  JUMPDEST 			* 
  POP 			contract C1 {\n    function f2...
```

## Environment

- Compiler version: 0.4.24 & 0.6.1
- Target EVM version (as per compiler settings): default
- Framework/IDE (e.g. Truffle or Remix): Remix

I think this needs a look.

### PR DESCRIPTION
**Pull request by mijovic**
Fixes #8171 

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### CODE REVIEW

File: libevmasm/Assembly.cpp [hunk 1]
```diff
@@ -217,65 +218,74 @@ string Assembly::toStringInHex(u256 _value)
 	return hexStr.str();
 }
 
-Json::Value Assembly::assemblyJSON(StringMap const& _sourceCodes) const
+Json::Value Assembly::assemblyJSON(map<string, unsigned> const& _indices) const
```
Comment (chriseth) [inline libevmasm/Assembly.cpp]:
```suggestion
Json::Value Assembly::assemblyJSON(map<string, unsigned> const& _sourceIndices) const
```
?

File: libevmasm/Assembly.cpp [hunk 2]
```diff
@@ -197,10 +197,11 @@ string Assembly::assemblyString(StringMap const& _sourceCodes) const
 	return tmp.str();
 }
 
-Json::Value Assembly::createJsonValue(string _name, int _begin, int _end, string _value, string _jumpType)
+Json::Value Assembly::createJsonValue(string _name, int _source, int _begin, int _end, string _value, string _jumpType)
```
Comment (chriseth) [inline libevmasm/Assembly.cpp]:
```suggestion
Json::Value Assembly::createJsonValue(string _name, int _begin, int _end, int _sourceIndex, string _value, string _jumpType)
```

Comment (chriseth) [inline libevmasm/Assembly.cpp:200]:
```suggestion
Json::Value Assembly::createJsonValue(string _name, int _sourceIndex, int _begin, int _end, string _value, string _jumpType)
```

File: libevmasm/Assembly.cpp [hunk 3]
```diff
@@ -217,65 +218,74 @@ string Assembly::toStringInHex(u256 _value)
...
-Json::Value Assembly::assemblyJSON(StringMap const& _sourceCodes) const
+Json::Value Assembly::assemblyJSON(map<string, unsigned> const& _indices) const
 {
 	Json::Value root;
 
 	Json::Value& collection = root[".code"] = Json::arrayValue;
 	for (AssemblyItem const& i: m_items)
 	{
+	    auto iter = _indices.find(i.location().source->name());
+	    int index = iter != _indices.end() ? iter->second : -1;
```
Comment (chriseth) [inline libevmasm/Assembly.cpp]:
```suggestion
	    int index = iter == _indices.end() ? -1 : iter->second;
```
(code with fewer negations is usually better to understand)

Comment (chriseth) [reply libevmasm/Assembly.cpp]: Also maybe `sourceIndex`?

File: libevmasm/Assembly.h [hunk 4]
```diff
@@ -133,7 +133,7 @@ class Assembly
 
 	/// Create a JSON representation of the assembly.
 	Json::Value assemblyJSON(
-		StringMap const& _sourceCodes = StringMap()
+        std::map<std::string, unsigned> const& _indices = std::map<std::string, unsigned>()
```
Comment (chriseth) [inline libevmasm/Assembly.h]: Please use tabs for indentation.

File: test/libsolidity/StandardCompiler.cpp [hunk 5]
```diff
@@ -394,27 +394,27 @@ BOOST_AUTO_TEST_CASE(basic_compilation)
...
-		"{\"begin\":0,\"end\":14,\"name\":\"POP\"},"
-		"{\"begin\":0,\"end\":14,\"name\":\"PUSH #[$]\",\"value\":\"0000000000000000000000000000000000000000000000000000000000000000\"},"
-		"{\"begin\":0,\"end\":14,\"name\":\"DUP1\"},"
-		"{\"begin\":0,\"end\":14,\"name\":\"PUSH [$]\",\"value\":\"0000000000000000000000000000000000000000000000000000000000000000\"},"
-		"{\"begin\":0,\"end\":14,\"name\":\"PUSH\",\"value\":\"0\"},"
-		"{\"begin\":0,\"end\":14,\"name\":\"CODECOPY\"},"
-		"{\"begin\":0,\"end\":14,\"name\":\"PUSH\",\"value\":\"0\"},"
-		"{\"begin\":0,\"end\":14,\"name\":\"RETURN\"}]"
+		"[{\"begin\":0,\"end\":14,\"name\":\"PUSH\",\"source\":0,\"value\":\"80\"},"
+        "{\"begin\":0,\"end\":14,\"name\":\"PUSH\",\"source\":0,\"value\":\"40\"},"
```
Comment (chriseth) [inline test/libsolidity/StandardCompiler.cpp]: Please use tabs.

**Comment by chriseth** [comment]
Please add a changelog entry.

File: libevmasm/Assembly.cpp (hunk 1 above)
Comment (mijovic) [reply libevmasm/Assembly.cpp]: Fixed as recommended.

File: libevmasm/Assembly.cpp (hunk 3 above)
Comment (mijovic) [reply libevmasm/Assembly.cpp]: Fixed as recommended.
Renamed index to sourceIndex as well as it is more descriptive name

File: test/libsolidity/StandardCompiler.cpp (hunk 5 above)
Comment (mijovic) [reply test/libsolidity/StandardCompiler.cpp]: Understood :)

**Comment by mijovic** [comment]
> Please add a changelog entry.

Hope explanation is clear enough in changelog

File: Changelog.md [hunk 6]
```diff
@@ -15,6 +15,7 @@ Compiler Features:
 Bugfixes:
  * Parser: Fix an internal error for ``abstract`` without ``contract``.
  * Type Checker: Make invalid calls to uncallable types fatal errors.
+ * Assembly: Added multi file sources to assembly json
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
 * Assembly: Added missing `source` field to legacy assembly json output to complete the source reference.
```

**Comment by mijovic** [comment]
Squashed to one commit and moved to another PR
https://github.com/ethereum/solidity/pull/8337

**Comment by chriseth** [comment]
Replaced by https://github.com/ethereum/solidity/pull/8337

**Pull request closed without merging**